### ISSUE TITLE
Layers are too restrictive with gotQueueFamilyPropertyCount for vkGetPhysicalDeviceSurfaceSupportKHR

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-LoaderAndValidationLayers#1720]
In validation layers there is this [condition](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/572ea79/layers/swapchain.cpp#L266) (simplified here):

`if( qfp_count && !qfp_is_null ) gotQueueFamilyPropertyCount = true;`

I think this is over-restrictive. E.g. this (IMO valid and reasonable) made up code would generate an validation error:

    bool fun( VkPhysicalDevice physicalDevice, VkSurfaceKHR surface ){
        uint32_t qfCount;
        vkGetPhysicalDeviceQueueFamilyProperties( physicalDevice, &qfCount, nullptr );

        if( qfCount > 0 ){
            VkBool32 isSupported;
            VkResult errorCode = vkGetPhysicalDeviceSurfaceSupportKHR(
                physicalDevice, queueFamily, surface, &isSupported
            );

            return isSupported;
        }
        
       return false;
    }

> ERROR: PhysicalDevice(0x17523E6A540): 24, Swapchain, vkGetPhysicalDeviceSurfaceSupportKHR() called before calling the vkGetPhysicalDeviceQueueFamilyProperties function.

<hr/>

I can PR this myself, given time.

<hr/>

Also the `phy_data->numOfQueueFamilies` seems to be rewritten every time on `vkGetPhysicalDeviceQueueFamilyProperties`.

And `phy_data->gotQueueFamilyPropertyCount` seems to be reset every `vkEnumeratePhysicalDevices`.

That gives an Error for many valid usages of Vulkan, and I suspect for some "reasonable" usages too,

**Comment by chrisforbes** [comment]
I agree, this is too aggressive. There are many enumeration-related cases that are like this-- we need to at least turn all the "we think this is poor practice but not invalid" checks into warnings.

**Comment by krOoze** [comment]
So I started thinking on how to fix this:

Basically, what I want is to determine when the application cannot know the queue family index it uses is valid or not (arguably that **should** still be an Error &mdash; there's no other way to validate queue family index. Though that check is brittle &mdash; someone may release an Extension also revealing queue family count and some outdated layer would issue a false-positive Error. But lets leave that to the philosophers.)

`vkGetPhysicalDeviceQueueFamilyProperties( physicalDevice, &qfCount, nullptr );` will always return the exact count.
`vkGetPhysicalDeviceQueueFamilyProperties( physicalDevice, &qfCount, someArray );` would return a count, but limited by the count the app provided.

So: my thinking is the `phy_data->numOfQueueFamilies` should be `max` of previous and current value. `phy_data->gotQueueFamilyPropertyCount` is a bit pointless, but I guess I could keep it (to make a clearer log message distinguishing it from 0 queue families count).
That sounds about right?

Implementing it I discovered the value is further reused here:
https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/572ea79/layers/swapchain.cpp#L1023

    pSurface->pQueueFamilyIndexSupport = 
    (VkBool32 *)malloc(pPhysicalDevice->numOfQueueFamilies * sizeof(VkBool32));

Let's ignore the hopeless C-ism for a minute. But I would appreciate if anybody who knows their way in this codebase would point me to the place where this is `free`d. Or is it a leak?
